Kernel/Bus/USB/USBConfiguration.cpp

```diff
@@ -15,61 +16,95 @@ namespace Kernel::USB {
+    auto const configuration_descriptor = TRY(stream.read_value<USBConfigurationDescriptor>());
+    if (configuration_descriptor.descriptor_header.length < sizeof(USBConfigurationDescriptor))
+        return EINVAL;
+
+    if (configuration_descriptor.total_length != m_descriptor.total_length)
+        return EINVAL;
+
+    USBInterface* current_interface = nullptr;
+
     TRY(m_interfaces.try_ensure_capacity(m_descriptor.number_of_interfaces));
-    for (auto interface = 0u; interface < m_descriptor.number_of_interfaces; interface++) {
-        endpoint_descriptors.ensure_capacity(interface_descriptor->number_of_endpoints);
-
-        if constexpr (USB_DEBUG) {
-            dbgln("Interface Descriptor {}", interface);
-            dbgln("interface_id: {:02x}", interface_descriptor->interface_id);
-            dbgln("alternate_setting: {:02x}", interface_descriptor->alternate_setting);
-            dbgln("number_of_endpoints: {:02x}", interface_descriptor->number_of_endpoints);
-            dbgln("interface_class_code: {:02x}", interface_descriptor->interface_class_code);
-            dbgln("interface_sub_class_code: {:02x}", interface_descriptor->interface_sub_class_code);
-            dbgln("interface_protocol: {:02x}", interface_descriptor->interface_protocol);
-            dbgln("interface_string_descriptor_index: {}", interface_descriptor->interface_string_descriptor_index);
-        }
 
-        // Get all the endpoint descriptors
-        for (auto endpoint = 0u; endpoint < interface_descriptor->number_of_endpoints; endpoint++) {
-            u8* raw_endpoint_descriptor_offset = reinterpret_cast<u8*>(interface_descriptor) + sizeof(USBInterfaceDescriptor) + (endpoint * sizeof(USBEndpointDescriptor));
+    auto read_descriptor = [&stream]<typename T>(USBDescriptorCommon const& header) -> ErrorOr<T> {
+        if (header.length < sizeof(T))
+            return EINVAL;
+
+        auto descriptor = TRY(stream.read_value<T>());
+
+        // Skip additional bytes.
+        TRY(stream.seek(header.length - sizeof(T), SeekMode::FromCurrentPosition));
 
-            // FIXME: It looks like HID descriptors come BEFORE the endpoint descriptors for a HID device, so we should load
-            // these too eventually.
-            // See here: https://www.usb.org/defined-class-codes
-            if (interface_descriptor->interface_class_code == USB_CLASS_HID)
-                raw_endpoint_descriptor_offset += sizeof(USBHIDDescriptor); // Skip the HID descriptor (this was worked out via buffer inspection)
+        return descriptor;
+    };
 
-            USBEndpointDescriptor* endpoint_descriptor = reinterpret_cast<USBEndpointDescriptor*>(raw_endpoint_descriptor_offset);
+    while (!stream.is_eof()) {
+        // Peek the descriptor header.
+        auto const descriptor_header = TRY(stream.read_value<USBDescriptorCommon>());
+        MUST(stream.seek(-sizeof(USBDescriptorCommon), SeekMode::FromCurrentPosition));
```

Comment:
It's a bit weird that we read the header, seek back, and then call a lambda that reads the header again. Might be nicer if we had a way to read just the tail (by using composition instead of inheritance in some strategic place). Fine for now, but could be nicer.

Reply:
I'm not really sure how I should refactor this to not read the header twice. We do need the headers in other places, so we can't just remove them from the structs.
Adding more structs for the tail would work but that would result in more struct definitions.